File: javaAlgo/src/test/java/leetcode/Task49Test.java
package leetcode;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import org.junit.jupiter.api.Test;

class Task49Test {

    private final Task49 task = new Task49();

    private static List<List<String>> canonical(List<List<String>> groups) {
        List<List<String>> sorted = new ArrayList<>();
        for (List<String> group : groups) {
            List<String> copy = new ArrayList<>(group);
            Collections.sort(copy);
            sorted.add(copy);
        }
        sorted.sort((a, b) -> String.join(",", a).compareTo(String.join(",", b)));
        return sorted;
    }

    @Test
    void example1() {
        String[] strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
        List<List<String>> expected = List.of(
                List.of("bat"),
                List.of("nat", "tan"),
                List.of("ate", "eat", "tea")
        );
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)));
    }

    @Test
    void example2() {
        String[] strs = {""};
        List<List<String>> expected = List.of(List.of(""));
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)));
    }

    @Test
    void example3() {
        String[] strs = {"a"};
        List<List<String>> expected = List.of(List.of("a"));
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)));
    }

    @Test
    void noAnagramsPresent() {
        String[] strs = {"abc", "def", "ghi"};
        List<List<String>> expected = List.of(
                List.of("abc"),
                List.of("def"),
                List.of("ghi")
        );
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)));
    }
}

File: kotlinAlgo/src/test/kotlin/leetcode/Task49Test.kt
package leetcode

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class Task49Test {
    private val task = Task49()

    private fun canonical(groups: List<List<String>>): List<List<String>> = groups.map { it.sorted() }.sortedBy { it.joinToString(",") }

    @Test
    fun example1() {
        val strs = arrayOf("eat", "tea", "tan", "ate", "nat", "bat")
        val expected = listOf(listOf("bat"), listOf("nat", "tan"), listOf("ate", "eat", "tea"))
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)))
    }

    @Test
    fun example2() {
        val strs = arrayOf("")
        val expected = listOf(listOf(""))
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)))
    }

    @Test
    fun example3() {
        val strs = arrayOf("a")
        val expected = listOf(listOf("a"))
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)))
    }

    @Test
    fun noAnagramsPresent() {
        val strs = arrayOf("abc", "def", "ghi")
        val expected = listOf(listOf("abc"), listOf("def"), listOf("ghi"))
        assertEquals(canonical(expected), canonical(task.groupAnagrams(strs)))
    }
}
